### ISSUE TITLE
Having problems creating a Custom Redactor

### ISSUE THREAD

**Issue by FirePT** [dotnet/extensions#4836]
### Description

Having problem creating a Custom Redactor. The value of the properties to be redacted are always appended by the TagName in the logs.
In the Redact method of the custom Redactor the value of the source parameter have the TagName appended at the end separated by the char ':'.

### Reproduction Steps

create Taxonomy class :

```
public static class DataTaxonomy
{
    public static string TaxonomyName { get; } = typeof(DataTaxonomy).FullName!;   

    public static DataClassification CreditCardData { get; } = new(TaxonomyName, nameof(CreditCardData));

}
```

create CreditCardData Attribute : 

 ```
 public class CreditCardDataAttribute : DataClassificationAttribute
  {
      public CreditCardDataAttribute() : base(DataTaxonomy.CreditCardData)
      {
      }
  }
```

create Redactor :

```
 internal class CreditCardRedactor : Redactor
 {
     public override int GetRedactedLength(ReadOnlySpan<char> input) => input.Length; 

     public override int Redact(ReadOnlySpan<char> source, Span<char> destination)
     {
         source.CopyTo(destination);
      
         return source.Length;
     }
 }
```

create Record :

```

public record CardDetailDto( [CreditCardData] string CardNumber, DateTime CreatedAt);

```

create Logger :

```
 public static partial class FeatureLogging
 {
     [LoggerMessage(100, LogLevel.Information, "Request received")]
     public static partial void LogRequest(this ILogger logger, [LogProperties] CardDetailDto card);
 }
```

### Expected behavior

```
{
  "Timestamp": "03:15:10 ",
  "EventId": 100,
  "LogLevel": "Information",
  "Category": "Card.Management.Api.Features.v1.GetCardDetailHandler",
  "Message": "Request received:",
  "State": {
    "Message": "Microsoft.Extensions.Logging.ExtendedLogger\u002BModernTagJoiner",
    "{OriginalFormat}": "Request received:",
    "card.CreatedAt": "12/23/2023 03:15:10",
    "card.CardNumber": "92400001"
  }
}
```

### Actual behavior

```
{
  "Timestamp": "03:15:10 ",
  "EventId": 100,
  "LogLevel": "Information",
  "Category": "Card.Management.Api.Features.v1.GetCardDetailHandler",
  "Message": "Request received:",
  "State": {
    "Message": "Microsoft.Extensions.Logging.ExtendedLogger\u002BModernTagJoiner",
    "{OriginalFormat}": "Request received:",
    "card.CreatedAt": "12/23/2023 03:15:10",
    "card.CardNumber": "92400001:card.CardNumber"
  }
}
```

### Known Workarounds

If I don't use the CreditCardDataAttribute the value of the CardNumber property is rendered correctly.

### Configuration

**.net 8**
**Microsoft.Extensions.Compliance.Redaction v8.0.0**
**Microsoft.Extensions.Telemetry v8.0.0**

Logging configuration :

```
 public static class Config
 {

     public static IServiceCollection ConfigureLogging(this IServiceCollection services)
     {
         services.AddLogging(opts =>
         {
             opts.EnableRedaction();
             opts.AddJsonConsole();
         });

         return services;
     }

     public static IServiceCollection ConfigureRedaction(this IServiceCollection services)
     {
         services.AddRedaction(opts =>
         {
             opts.SetRedactor<CreditCardRedactor>(new DataClassificationSet(DataTaxonomy.CreditCardData));
         });

         return services;
     }

 }
```

**Comment by geeknoid** [comment]
First of all, thank you very much for the truly excellent bug report. It took me 5 minutes to create a repro, nice.

Turns out this is a feature, not a bug. See <https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerredactionoptions.applydiscriminator?view=dotnet-plat-ext-8.0#microsoft-extensions-logging-loggerredactionoptions-applydiscriminator> for details on the option that controls this behavior.

The common use of redactors is to hash data. If some PII is used in a log statement, the PII is hashed and so when you look at the logs you can't recover the PII. Hashing is superior to merely removing the data as a whole, since it makes it possible to use correlation between multiple log statements. For example, you might not know what a user's name is by looking at the logs, but you can tell it's the same user that did actions 1, 2, and 3 over time, since the hash codes match. OK, all good so far.

During security audits of the feature, our security guys got worried that you could potentially combine multiple pieces of state together to reverse engineer the contents of the log and recover the user's name, thus defeating the whole reason for redaction. The solution we came up with was this discriminator idea. It blends in the name of the property to the state to be redacted. This means that the hashed values will be different for two properties even if their data was the same. 

In other words:

```
class Example
{
    public string FIeld1 { get; } = "ABC";
    public string Field2 { get; } = "ABC";
}
```

if I log the above using [LogProperties] and a hashing redactor, the resulting hashed values emitted in the logs for Field1 and Field2 will be different, even though the input data was the same (namely "ABC"). And this means that you cannot correlate data in the log if the property names don't match. And this is the key to prevent the problem.

Anyway, you can turn the option off and the extra data will not be added to the state.

**Comment by FirePT** [comment]
I @geeknoid , thanks for the feedback. I understand the reasoning of the added discriminator and the added security it provides  but I can't get past the fact that the current implementation either have the added security or the freedom to create my custom redactors. When defining my custom one I didn't disable the "applydiscriminator" because I would loose the added security on the out of the box redactors you already provide, so I kept it, expecting to have full control over the transformation between the **source** and **destination**.

My requirement is to transform a credit card number in a string with the same length but with the chars replaced by the char '-' other than the last three. (378282246310005 => ------------005) 


simplified version
```
 public override int Redact(ReadOnlySpan<char> source, Span<char> destination)
 {
     source.CopyTo(destination);

     var splitPos = source.IndexOf(':');

     destination[..(splitPos - 3)].Fill('-');
     }
     destination = destination[..splitPos];

     return splitPos ;
 }
```

Inspecting destination before the return, I can see that the span is assigned with the correct string, in the correct format. However, you still add the discriminator afterwards. Because of this, I don't really have control over the output of my custom redactor.

Am I missing something here ? I was expecting not to have to choose between security and a custom implementation.

**Comment by geeknoid** [comment]
Indeed, the discriminator is applied globally independent of the redactor's output. This is necessary since redactor's are aware of what they are redacting, they just convert one string to another string, ignoring the context these strings came from.

We could potentially extend the model in one of two ways:

* Add control of discriminator use per data classification.
* Add control of discriminator use per redactor type.

Do you have a preference for which approach would work best for you?

**Comment by FirePT** [comment]
For me, the most elegant approach would involve _**per redactor type**_. Similar to how we override the **GetRedactedLength** method, we could also override a **Boolean** return method. This method would control the usage of the discriminator, defaulting to **True** in the base, thereby enforcing security by default.

Something like :

`public override bool ApplyDiscriminator() => return false;`

 Additionally, this would streamline the API configuration, as it currently it's the only available option for the **EnableRedaction** configuration method and could go away (in this case a breaking change :( ).
